tests/test_crf.py

```diff
@@ -116,7 +119,10 @@ def test_works_with_mask(self):
             denominator = math.log(sum(math.exp(s) for s in all_scores))
             manual_llh += numerator - denominator
 
-        assert llh.item() == approx(manual_llh)
+        if HAS_ASSERT_CLOSE:
+            torch.testing.assert_close(llh, manual_llh, atol=1e-12, rtol=1e-6)
+        else:
+            torch.testing.assert_allclose(llh, manual_llh, atol=1e-12, rtol=1e-6)
```

Comment:
Can you explain what `assert_close` and `assert_allclose` do that `approx` doesn't?

Reply:
The problem is not in approx itself, but it requires converting from torch tensor to python scalar, likely through numpy. I started to see `RuntimeError: Can't call numpy() on Tensor that requires grad. Use tensor.detach().numpy() instead.` when using torch 1.4. Instead of adding `.detach().numpy().item()` all over the place, it feels simpler and more readable to use dedicated testing utilities provided by torch itself.

Reply:
I'll change the commit message to make it more clear. The error mentioned in the commit seemed to only happen in my local environment.

Reply:
I see. I agree that adding `.detach().numpy().item()` all over the place is less readable. However, I believe Pytest's `approx` has a nicely formatted error message when the test fails, which is also important in terms of readability. Does PyTorch's `assert_close` have it too? Can you please compare them and share it here?

Reply:
Yes. The following code

```python
def test_approx():
    a = 1/3
    b = 0.3
    assert a == approx(b)

def test_torch_assert_close():
    a = torch.tensor(1/3)
    b = torch.tensor(0.3)
    torch.testing.assert_close(b, a)
```

gives
![image](https://github.com/kmkurn/pytorch-crf/assets/46868116/b289f8b7-b9f6-42b8-b902-77175c798ca9)

Reply:
Thanks! The PyTorch one is readable too, which is good. Now, to improve readability further, can you please move the calling logic to a separate function? Something like:

```python
def assert_close(value, /, expected):
    fn_name = "assert_close" if Version(torch.__version__) >= Version("1.9.0") else "assert_allclose"
    assert_fn = getattr(torch.testing, fn_name)
    return assert_fn(value, expected, atol=1e-12, rtol=1e-6)
```

Then you can use `assert_close` in all the tests to replace `approx`.

Also, if `approx` is no longer used anywhere, please delete its import statement (line 5).

Reply:
The requested changes have been added. I did not use string literals for function names because it makes editor hovering and completions impossible.